### PR TITLE
#948 Add option "Start in system tray" in "Settings" window

### LINKED ISSUE #948: Add option "Start in system tray" in "Settings" window
As explained in issues #407 and #422, there is already a command line option to start Telegram minimized in the system tray. However, it doesn't work well when using a window manager such as KDE. The problem I have found is that window managers (KDE) remember the applications minimized in the system tray and start them up every time you login. The problem is that it doesn't pass the -startintray option, resulting in Telegram starting up in normal way (i.e., non-minimized). 

By having a new "Start in system tray" option in the "Settings" window, the user could check or uncheck easily the desired behavior. It also removes the need of launching Telegram from command line to get it minimized. Furthermore, the only way to achieve this goal currently without recurring to command line is by modifying the Telegram script (/usr/bin/telegram) which not only is a bad practice, but it also gets overridden every time the Telegram package is updated. If this is not enough, it sets the same behavior for all users of that PC which could not be the same necessarily as well.

To sum up, the feature is already implemented and adding this new option would greatly improve the usability with (IMHO) really little additional effort.

Thanks a lot! ;-)

### PR DESCRIPTION
**Pull request by Musikolo**
Implementation of feature request closes #948. Built and tested against latest upstream dev branch. 

If anything needs to be improved, please just let me know.

### CODE REVIEW

File: Telegram/Resources/langs/lang_es.strings [hunk 1]
```diff
@@ -315,7 +315,8 @@ Copyright (c) 2014-2017 John Preston, https://desktop.telegram.org
 "lng_settings_workmode_tray" = "Mostrar el icono";
 "lng_settings_workmode_window" = "Mostrar icono de la barra de tareas";
 "lng_settings_auto_start" = "Abrir Telegram cuando el sistema inicie";
-"lng_settings_start_min" = "Abrir minimizada";
+"lng_settings_start_min" = "Iniciar minimizado";
+"lng_settings_start_in_tray" = "Iniciar en la bandeja de sistema";
```
Comment (stek29) [inline Telegram/Resources/langs/lang_es.strings]: You should only add English translation

Comment (Musikolo) [reply Telegram/Resources/langs/lang_es.strings]: Does this mean you want me to remove Spanish translations? 

I just want to confirm.

Thank you.

**Comment by Musikolo** [comment]
For some reason CI test 5 failed while building 3rd party dependencies (Qt & range-v3) as you can see at https://travis-ci.org/telegramdesktop/tdesktop/builds/323815838

Previous built was completely successful though: https://travis-ci.org/telegramdesktop/tdesktop/builds/323729076

I don't think this error is related to my last code change, but I'm missing something, please, just let me know.

Thank you!

**Comment by auchri** [comment]
Everything is ok with the builds:

screenshot_20180101-205535

**Comment by john-preston** [comment]
@auchri I've restarted it.

File: Telegram/Resources/langs/lang_es.strings (hunk 1 above)
Comment (auchri) [reply Telegram/Resources/langs/lang_es.strings]: yes.

**Comment by Musikolo** [comment]
@auchri, the changes you are requested are done.

Please, let me know if anything else is needed.

Thank you!

**Comment by john-preston** [comment]
1. This startInTray option currently is controlled by "-startintray" command line argument. After your changes the command line argument is ignored and the value is always read from settings.

This is not good, because "-startintray" argument is passed from "launch when system starts" shortcut on Windows and it is controlled by a separate "launch minimized" setting, which will be just ignored in case your changes are added.

2. This startInTray option and "-startintray" command line argument were added for this specific case - for "launch when system starts" + "launch minimized" on Windows.

It works there, but it was not tested and is known not to work in other systems and configurations, so it is a bad idea to give users a shiny checkbox in settings for something that perhaps doesn't work for them or works bad. See #512 #2123 #768 #3508 #3391 #1326 #422 #407 #511 #1820

With all this said I remind you that "we do not merge any pull requests that introduce any new user interface elements" was added to our contributing guidelines for a reason :(

**Comment by Musikolo** [comment]
@john-preston, thank for your feedback.

I understand your concerns about the possible unwanted side effects my implementation could cause. Being honest, I have only tested the code in my Arch Linux box with KDE. Per the comments we have at #948, many people confirmed the `-startintray` command line option works for the Linux platform overall, plus I can confirm it works for me too. 

Furthermore, this feature requested in the issue #948 seems to be claimed or wanted by a relatively big number of users over a long period of time (~ 2.5 years). So, I would like to know your advise on how we can find a way of merging both worlds, i.e., do not alter the existing functionality while incorporating a settings option that improves the user experience. 

- Is there any possible approach I could follow that you feel comfortable with?
- Would it be an option to make the code I implemented Linux-platform specific (i.e., only available for the Linux platform)?

Looking forward to your feedback.

Thank you so much for your time! ;-)

**Comment by Aokromes** [comment]
@Musikolo  if you change:

Implementation of feature request #948.

To

Implementation of feature request closes #948.
It will autocloses that ticket :)

**Comment by Musikolo** [comment]
@Aokromes I don't want to close the issue #948 until there is an agreed solution to it. My PR hasn't been accepted yet, so it cannot be considered (yet) a solution for it. I'm waiting for @john-preston and/or other developers to provided me with guidance on how I could help it get such an issue solved.

Thanks for your suggestion!

**Comment by Aokromes** [comment]
I mean, when this pr is merged with the suggested text it will close that issue :)

**Comment by Musikolo** [comment]
@Aokromes, OK, I see now. It should be done now. ;)

I hope this feature is merged in the near future. 

Thank you!

**Comment by Musikolo** [comment]
@john-preston / @auchri, I've been using Telegram Desktop with my patch seamlessly for over a month.  Following your last comments, the only scenario that changes the existing behavior requires the next steps:

1. Start Telegram Desktop.
1. Check the option "_Launch in system tray_" (new checkbox).
1. Exit Telegram Desktop.
1. Start Telegram from command line without the switch "`-startinsystemtray`".

In this scenario, the application shows in the system tray, but it does so because you manually selected an option to do exactly that, start in the system tray. If you want the previous behavior back, all you need to do is to uncheck such an option (_Launch in system tray_).

From my point of view, there is nothing wrong in what the patch does. I believe there are no other behavioral changes beyond the scenario described above. After 2.5 years waiting for a solution, I (and a few more #948 ) would truly like my patch incorporated into the upstream code base. If you don't feel comfortable with the code changes, I'm willing to update it the way consider necessary, provided I get some guidance on what's expected.

I appreciate your time and good job.

Looking forward to your reply!

**Comment by stek29** [comment]
Maybe you'd implement changes @auchri has requested while waiting at least?

**Comment by Musikolo** [comment]
@stek29, the only changes @auchri requested me [were implemented](https://github.com/telegramdesktop/tdesktop/pull/4268#issuecomment-354920857) a month ago in [this commit](https://github.com/telegramdesktop/tdesktop/pull/4268/commits/b0a0348d75f437cf3280e000e45819b1eed31ea4). However, if I've missed anything, I would be more than happy to get it done asap. 

Please, let me know, if this is the case.

Thank you!

**Comment by stek29** [comment]
@Musikolo oh, sorry, I've missed that somehow.

**Comment by abixadamj** [comment]
Strange behaviour on Mint 18.2 64bit
I have Installed 1.2.6 version yesterday. Everything works perfectly. I have checked 'Show tray icon' and Telegram runs OK. I have added Telegram to Autostart in Mint. Today I have started computer, and Telegram opened in normal windows, without system tray icon.
When I go to Preferences, uncheck 'Show tray icon" and then check it again, icon apperad in tray.

**Comment by Musikolo** [comment]
As I mentioned 2 months ago, if there is anything I can do to meet your requirements and get this PR merged, please, let me know.

Will I ever get any reply or feedback? :-/

**Comment by john-preston** [comment]
As I mentioned earlier a minimum requirement is that the current workflow on other platforms won't be broken. For that you must always start in tray in case "-startintray" argument was passed, no matter what is saved in the settings. This could be done if "-startintray" and your setting flag were held in different variables, not in the same one.

Also I think you need to merge the submodules as well (maybe run "git submodule update" after "git merge upstream/dev" or smth like that), because currently your patch show changes in submodules, which there should not be.

**Comment by Musikolo** [comment]
Closing this PR since it's being replaced by new [PR #4782](https://github.com/telegramdesktop/tdesktop/pull/4782).

**Pull request closed without merging**